li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaInstrumentedConsumerImpl.java

```diff
@@ -0,0 +1,583 @@
+import com.linkedin.mario.simpleclient.SimpleClientState;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.clients.consumer.OffsetCommitCallback;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class LiKafkaInstrumentedConsumerImpl<K, V> implements Consumer<K, V>, EventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LiKafkaInstrumentedConsumerImpl.class);
+
+  private final ReadWriteLock delegateLock = new ReentrantReadWriteLock();
+  private final Properties baseConfig;
+  private final Map<String, String> translatedBaseConfig;
+  private final Function<Map<String, String>, Consumer<K, V>> consumerFactory;
+  private final String mdsUrl;
+  private final CountDownLatch initialConnectionLatch = new CountDownLatch(1);
+  private final MetricsProxy metricsProxy = new MetricsProxy();
+  private volatile Long closedAt = null;
+  private volatile Map<String, String> configOverrides;
+  private volatile Consumer<K, V> delegate;
+  private SimpleClient mdsClient;
+
+  private volatile Collection<String> subscribedTopics = null;
+  private volatile ConsumerRebalanceListener rebalanceListener = null;
+  private volatile Collection<TopicPartition> assignedPartitions = null;
+  private volatile Pattern subscriptionPattern = null;
+
+  public LiKafkaInstrumentedConsumerImpl(
+      Properties baseConfig,
+      Function<Map<String, String>, Consumer<K, V>> consumerFactory,
+      String mdsUrl
+  ) {
+    List<String> conversionIssues = new ArrayList<>(1);
+    this.baseConfig = baseConfig;
+    this.translatedBaseConfig = LiKafkaClientsUtils.toStringMap(baseConfig, conversionIssues);
+    this.consumerFactory = consumerFactory;
+    this.mdsUrl = mdsUrl;
+
+    if (!conversionIssues.isEmpty()) {
+      StringJoiner csv = new StringJoiner(", ");
+      conversionIssues.forEach(csv::add);
+      LOG.error("issues translating consumer config to strings: {}", csv);
+    }
+
+    mdsClient = new SimpleClient(
```

Comment:
Suggestion: Is it worth adding a handle in `SimpleClient` to wait on connection to complete / timeout , instead of repeating this connection handling logic in both the producer and consumer implementations? 
Fyi, I haven't looked at `SimpleClient` . So, my comment may be incorrect.

Reply:
i tried not to force users of SimpleClient into a "blocking constructor" pattern. its more flexible that way, at the cost of repeating the countdown logic (which isnt all that much, i think)